agent/consul/util.go

```diff
@@ -273,89 +273,163 @@ func runtimeStats() map[string]string {
+	state := serversMeetRequirementsState{meetsRequirements: meetsRequirements, found: false, ok: true}
+
+	provider.CheckServers(datacenter, state.update)
+
+	return state.ok, state.found
 }
 
 // ServersMeetMinimumVersion returns whether the given alive servers from a particular
-// datacenter are at least on the given Consul version. This requires at least 1 alive server in the DC
-func ServersInDCMeetMinimumVersion(members []serf.Member, datacenter string, minVersion *version.Version) (bool, bool) {
-	found := false
-	ok := ServersMeetRequirements(members, func(srv *metadata.Server) bool {
-		if srv.Status != serf.StatusAlive || srv.Datacenter != datacenter {
-			return true
+// datacenter are at least on the given Consul version. This also returns whether any
+// alive or failed servers are known in that datacenter (ignoring left and leaving ones)
+func ServersInDCMeetMinimumVersion(provider checkServersProvider, datacenter string, minVersion *version.Version) (ok bool, found bool) {
+	return ServersInDCMeetRequirements(provider, datacenter, func(srv *metadata.Server) (bool, bool) {
+		if srv.Status != serf.StatusAlive && srv.Status != serf.StatusFailed {
+			// filter out the left servers as those should not be factored into our requirements
+			return true, true
 		}
 
-		found = true
-		return !srv.Build.LessThan(minVersion)
+		return !srv.Build.LessThan(minVersion), false
 	})
+}
 
-	return ok, found
+// CheckServers implements the checkServersProvider interface for the Server
+func (s *Server) CheckServers(datacenter string, fn func(*metadata.Server) bool) {
+	if datacenter == s.config.Datacenter {
+		// use the ServerLookup type for the local DC
+		s.serverLookup.CheckServers(fn)
+	} else {
+		// use the router for all non-local DCs
+		s.router.CheckServers(datacenter, fn)
+	}
 }
 
-// ServersMeetRequirements returns whether the given server members meet the requirements as defined by the
-// callback function
-func ServersMeetRequirements(members []serf.Member, meetsRequirements func(*metadata.Server) bool) bool {
-	for _, member := range members {
-		if valid, parts := metadata.IsConsulServer(member); valid {
-			if !meetsRequirements(parts) {
-				return false
-			}
+type serversACLMode struct {
+	// leader is the address of the leader
+	leader string
+
+	// mode indicates the overall ACL mode of the servers
+	mode structs.ACLMode
+
+	// leaderMode is the ACL mode of the leader server
+	leaderMode structs.ACLMode
+
+	// indicates that at least one server was processed
+	found bool
+}
+
+func (s *serversACLMode) init(leader string) {
+	s.leader = leader
+	s.mode = structs.ACLModeEnabled
+	s.leaderMode = structs.ACLModeUnknown
+	s.found = false
+}
+
+func (s *serversACLMode) update(srv *metadata.Server) bool {
+	if srv.Status != serf.StatusAlive && srv.Status != serf.StatusFailed {
+		// they are left or something so regardless we treat these servers as meeting
+		// the version requirement
+		return true
```

Comment:
🤔 I know this isn't changing behaviour but I wonder if we need to check for quorum here somewhere?

It's an edge case but if the leader got partitioned right after it won election and started checking ACL version etc. I think this code could allow it to continue with the ACL upgrade even if all other servers in DC are on an older version.

I guess if it did do so then any new state it tried to write to FSM would fail and other servers if they saw the append messages afterwards should probably ignore them before they try to parse them because the term would be too old. But I guess with timing it's just about possible that serf could report all but the current server as failed, we decide "all non failed servers are updated" then write new ACL stuff to FSM and the partiion heals enough that it commits before heartbeats are missed and so we've managed to write bad state onto older servers.

I don't know if we need to fix that here - it's such an extreme edge case during an upgrade which is a relatively rare event that it feels overwhelmingly unlikely to happen but should we note it somewhere?

Reply:
In short, I think all the edge cases are accounted for and the system should operate correctly even if there is a network partition.

Things only transition if all the alive and failed servers in the DC can support new ACLs. Left/Leaving servers are not considered. We consider failed servers as they could still come back online but left ones we assume will not be coming back.

So we only ever initialize new ACL resources when all servers can support new ACLs (whether they have "transitioned" or not doesn't really matter within a single DC). In fact under normal conditions, the global-management policy could already be applied through raft prior to any of the non-leader servers in the DC having noticed that the leader transitioned and marked themselves as transitioned. RPCs to those untransitioned servers for new ACL endpoints will return an error until they have picked up on the leader transitioning but the internal raft apply will still succeed.

So basically, a partition could cause the raft apply of the global-management policy etc. to fail but thats pretty much the same for any writes you might perform and then lose leadership. When the partition gets restored either the other server nodes maintained a leader during the partition or a quorum could not be had. In the first case that new leader would have transitioned itself and injected the ACL resources and when the partitioned node comes back it will get those resources via a snapshot or raft logs. In the second case, once the partition is restored a leader will get elected and either blindly initialize ACL resources if it had already transitioned but failed to do the raft writes, or do the check to ensure that it can (and then still initialize because all the servers support new ACLs).